arch/x86_64/core/x86_64-hw.h

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef _X86_64_HW_H
+#define _X86_64_HW_H
+
+/*
+ * Struct declarations and helper inlines for core x86_64 hardware
+ * functionality.  Anything related to ioports, CR's MSR's, I/L/GDTs,
+ * PTEs or (IO-)APICs can be found here.  Note that because this
+ * header is included in limited stub contexts, it should include
+ * declarations and inlines only: no data definitions, even extern
+ * ones!
+ */
+
+static inline unsigned long eflags(void)
+{
+	int eflags;
+
+	__asm__ volatile("pushfq; pop %%rax" : "=a"(eflags));
+	return eflags;
+}
+
+/* PAE page table record.  Note that "addr" is aligned naturally as an
+ * address, but of course must be masked to change only significant
+ * bits (which depend on whether it's storing a 4k, 2M or 1G memory
+ * block) so as to not clobber the bitfields (remember "negative"
+ * addresses must mask off the top bits too!).  The natural idiom is
+ * to assign addr first, then write the bitfields.
+ */
+struct pte64 {
```

Comment:
I think we already have 64-bit page table struct definitions in arch/x86/include/mmustructs.h see x86_mmu_pae_pte

Can we perhaps have both x86 and x86_64 ports use this header?

Reply:
Yeah, that for sure is the goal.  There is no attempt here to support userspace, I just needed the page tables set up so that I could enter long mode.  Enabling that and unifying the implementations (honestly I like the way the struct-with-bitfield style looks here when used in code) is for sure the next step.

Right now though that's a lot of code sharing across modes -- I need these defined in 32 bit code during bootstrapping, and didn't want to play with all the cmake madness to get the includes and paths to be cross-arch hygienic.